pkg/pillar/zedcloud/deferred.go

```diff
@@ -24,97 +27,149 @@ import (
+func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.Duration, sendOne bool) bool {
 
-	return zedcloudCtx.deferredCtx.handleDeferred(zedcloudCtx.log, event, spacing)
+	return zedcloudCtx.deferredCtx.handleDeferred(zedcloudCtx.log, event, spacing, sendOne)
 }
 
 func (ctx *DeferredContext) handleDeferred(log *base.LogObject, event time.Time,
-	spacing time.Duration) bool {
+	spacing time.Duration, sendOne bool) bool {
+	ctx.lock.Lock()
+	defer ctx.lock.Unlock()
+
+	if len(ctx.deferredItems) == 0 {
+		return true
+	}
 
-	log.Functionf("HandleDeferred(%v, %v) map %d\n",
+	log.Functionf("HandleDeferred(%v, %v) items %d",
 		event, spacing, len(ctx.deferredItems))
+
+	//sort items by time the entry was added/created in ascending order
+	sort.Slice(ctx.deferredItems, func(i, j int) bool {
+		t1 := ctx.deferredItems[i].lastModifiedTime
+		t2 := ctx.deferredItems[j].lastModifiedTime
+		return t1.Before(t2)
+	})
+
 	iteration := 0 // Do some load spreading
-	for key, l := range ctx.deferredItems {
-		log.Functionf("Trying to send for %s items %d\n", key, len(l.list))
-		failed := false
-		for i, item := range l.list {
-			if item.buf == nil {
-				continue
-			}
-			if item.buf.Len() == 0 {
-				log.Errorf("Zero length defered item for %s",
-					key)
-				continue
-			}
-			log.Functionf("Trying to send for %s item %d data size %d\n",
-				key, i, item.size)
-			resp, _, _, err := SendOnAllIntf(item.zedcloudCtx, item.url,
-				item.size, item.buf, iteration, item.bailOnHTTPErr)
-			if item.bailOnHTTPErr && resp != nil &&
-				resp.StatusCode >= 400 && resp.StatusCode < 600 {
-				log.Functionf("HandleDeferred: for %s ignore code %d\n",
-					key, resp.StatusCode)
-			} else if err != nil {
-				log.Functionf("HandleDeferred: for %s failed %s\n",
-					key, err)
-				failed = true
-				break
+	done := false
+	for _, f := range ctx.priorityCheckFunctions {
+		for ind := 0; ind < len(ctx.deferredItems); ind++ {
+			key := ctx.deferredItems[ind].key
+			item := ctx.deferredItems[ind]
+			if f(item.itemType) {
+				log.Functionf("handleDeferred: Trying to send for %s", key)
+				if item.buf == nil {
+					continue
+				}
+				if item.buf.Len() == 0 {
+					log.Functionf("handleDeferred: Zero length deferred item for %s",
+						key)
+					continue
+				}
+				resp, _, _, err := SendOnAllIntf(ctx.zedcloudCtx, item.url,
+					item.size, item.buf, iteration, item.bailOnHTTPErr)
+				if item.bailOnHTTPErr && resp != nil &&
+					resp.StatusCode >= 400 && resp.StatusCode < 600 {
+					log.Functionf("handleDeferred: for %s ignore code %d",
+						key, resp.StatusCode)
+				} else if err != nil {
+					log.Functionf("handleDeferred: for %s failed %s",
+						key, err)
+					done = true
+					break
+				}
+				if ctx.sentHandler != nil {
+					f := *ctx.sentHandler
+					f(item.itemType, item.buf)
+				}
+				item.buf = nil
+				ctx.deferredItems = append(ctx.deferredItems[:ind], ctx.deferredItems[ind+1:]...)
```

Comment:
One thing to watch for with slices and append is increasing and unbounded memory wastage.
If not careful the underlying array (i.e., cap(ctx.deferredItems) can grow larger and larger even thous its len does not.). Don't know if this particular pattern ends up allocating a new array or potentially growing the existing one.

Reply:
I modify the logic with cleanup after all sendings